standard/lexical-structure.md

```diff
@@ -1488,12 +1488,12 @@ fragment PP_Line_Indicator
     ;
     
 fragment PP_Compilation_Unit_Name
-    : '"' PP_Compilation_Unit_Name_Character+ '"'
+    : '"' PP_Compilation_Unit_Name_Character* '"'
```

Comment:
We should check this is *by design* before making this change. Regardless of the answer there is probably some work to do:

- §14.2 *Compilation units* defines a compilation unit, the definition does *not* include them having names… or being files… However the example (non-normative) uses two files `A.cs` and `B.cs` and refers to them as two compilation units.
- §22.5.6.3 *The CallerFilePath attribute* which provides the file path (which is implementation-dependent) states “The file path may be affected by `#line` directives ([§6.5.8]”.
- Here in §6.5.8 the `#line` allows the setting of the “compilation unit name”
- So:
  - §6.5.8 states compilation units have names, which is omitted in §14.2; and
  - §22.5.6.3 tells us that the name is the file path, but leaves what that is implementation-dependent
 
What already exists is overly clear and this change seeks to allow the compilation unit name to be the empty string, which is probably not a valid implementation-dependent path on any implementation… So if this
observed compiler behaviour is by design then it surely needs to have a defined meaning in the Standard.

If this change is to be made, and even if not, this all needs to tided up – either in this PR or spin it all off into a new one.

---

I might ask what the intended use of an empty file path/compilation unit name is but I might know – it was requested by the NSA so that the file names in NSA distributed software is not leaked and so endanger National Security! I’m only partially joking here, but that’s a story for another time ;-)

Reply:
@BillWagner told me that allowing an empty string was indeed a conscious decision, so I propose keeping that edit.

Reply:
@Nigel-Ecma 

[§6.1 Programs](lexical-structure.md#61-programs) states:

> A C# ***program*** consists of one or more source files, known formally as ***compilation units*** ([§14.2](namespaces.md#142-compilation-units)). Although a compilation unit might have a one-to-one correspondence with a file in a file system, such correspondence is not required.

I propose appending to this, the following:

> … As such, the accepted spelling of a compilation unit name, and its mapping, if any, to a filename is outside the scope of this specification.

I'm deliberately avoiding using any of the following terms:

- behavior, implementation-defined – unspecified behavior where each implementation documents how the choice is made
- behavior, undefined – behavior, upon use of a non-portable or erroneous construct or of erroneous data, for which this specification imposes no requirements
- behavior, unspecified – behavior where this specification provides two or more possibilities and imposes no further requirements on which is chosen in any instance

Reply:
This issue is now raised in #1233.